census_view.cpp

```diff
@@ -932,13 +936,839 @@ inline std::vector<std::string> const& CensusViewDataViewModel::all_headers() co
+            }
+        ,
+            { get_impl<table_date_converter>().type()
+            ,&get_impl<table_date_converter>()
+            }
+        ,
+        };
+    return all;
+}
+
+table_type_converter const&
+table_type_converter::get_by_value(any_member<Input> const& value)
+{
+    if(exact_cast<mce_yes_or_no>(value))
+        {
+        return get_impl<table_bool_converter>();
+        }
+    else if(exact_cast<datum_string>(value))
+        {
+        return get_impl<table_string_converter>();
+        }
+    else if(is_reconstitutable_as<datum_sequence>(value))
+        {
+        return get_impl<table_sequence_converter>();
+        }
+    else if(is_reconstitutable_as<mc_enum_base  >(value))
+        {
+        return get_impl<table_enum_converter>();
+        }
+    else if(is_reconstitutable_as<tn_range_base >(value))
+        {
+        tn_range_base const* as_range = member_cast<tn_range_base>(value);
+        if(typeid(int) == as_range->value_type())
+            {
+            return get_impl<table_int_range_converter>();
+            }
+        else if(typeid(double) == as_range->value_type())
+            {
+            return get_impl<table_double_range_converter>();
+            }
+        else if(typeid(calendar_date) == as_range->value_type())
+            {
+            return get_impl<table_date_converter>();
+            }
+        else
+            {
+            // Fall through to warn and treat datum as string.
+            }
+        }
+    else
+        {
+        warning()
+            << "Type '"
+            << lmi::TypeInfo(value.type())
+            << "' not recognized. Please report this anomaly."
+            << LMI_FLUSH
+            ;
+        // Fall through to treat datum as string.
+        }
+
+    return get_impl<table_string_converter>();
+}
+
+template<typename T>
+table_type_converter const& table_type_converter::get_impl()
+{
+    static T singleton;
+    return singleton;
+}
+
+} // Unnamed namespace.
+
+/// Interface to the data for wxGrid.
+
+class CensusViewTable : public wxGridTableBase
+{
+  public:
+    // Cell serial number: always shown in first column.
+    static int const Col_CellNum = 0;
+
+    explicit CensusViewTable(CensusGridView& view)
+        :view_ {view}
+    {
+    }
+
+    // return the number of rows and columns in this table.
+    int GetNumberRows() override;
+    int GetNumberCols() override;
+
+    wxString GetValue(int row, int col) override;
+    void SetValue(int row, int col, wxString const& value) override;
+
+    // Data type determination.
+    wxString GetTypeName(int row, int col) override;
+
+    // Override only used rows/cols handling functions.
+    bool AppendRows(size_t numRows) override;
+    bool DeleteRows(size_t pos, size_t numRows) override;
+    bool AppendCols(size_t numCols) override;
+    bool DeleteCols(size_t pos, size_t numCols) override;
+
+    wxString GetColLabelValue(int col) override;
+
+    void SetFirstColumnReadOnly();
```

Comment:
It looks like this ought to be called `make_cell_number_column_read_only()` because:

1. It's lmi-specific function, so it should follow lmi naming convention, not wx one.
1. The important thing is that this column contains the cell number, not that it's the first one.
1. "Make read-only" reads better than "set read-only".